### ISSUE TITLE
Faster loading of Tab Stash UI

### ISSUE THREAD

**Issue by joshgold22** [josh-berry/tab-stash#97]
I often don't need to examine unstashed tabs when I open the Tab Stash view (either sidebar or in a tab.)  Could the Unstashed Tabs group be added only on a button press after opening?

My most important motivation is performance.   I believe they are the cause of a very long delay opening the view when many tabs and/or windows are open, and even was concerned that I'd permanently lost tabs for a research project.

Other thoughts: For stashing, I personally use multiple selection (via Firefox's built-in mechanisms, or my own preferred extensions, depending what I'm doing) and the context menu, tool button, or keyboard shortcut.  Also, the Unstashed group does take up space from the groups I want to work with, when I'm not going to the view to add more to a group.

### Preferred Solution(s)

A button at the top of the stashes "view" to show the current tabs.  (Perhaps there could be variants for all tabs vs. current window.

There could be a preference to show Unstashed Tabs when the view is opened.  I'd suggest default to be false, since the performance made me think the extension was buggy and gave me concern I'd lost tabs permanently.

### Alternative Solution(s)

A preference option to include Unstashed Tabs in the Stashes views or not.  (This would be less flexible since it may be useful on occasion.)

### Additional Context

<hr>
<em>Readers: If you are also interested in seeing this feature be developed, please vote for it by giving the ORIGINAL POST a thumbs-up using the :smiley: button.  You are welcome to leave comments and discuss the feature request, but "Me too!" comments are not counted by the voting system.</em>

**Comment by josh-berry** [comment]
Hi, you should be able to collapse the "Unstashed Tabs" list by clicking on the arrow on the right which appears when you hover over the list, and Tab Stash should remember that you collapsed it. Does this address your concern?

<img width="329" alt="Screen Shot 2020-11-14 at 17 25 35" src="https://user-images.githubusercontent.com/1203893/99160685-a3385b80-269e-11eb-8271-411e2fc48139.png">

Regarding the performance issue, the unstashed tabs aren't the problem, it's loading the stash itself from your Firefox bookmarks that takes most of the time—this is the single biggest operation that happens during page load. I have some ideas for how to improve it, but I just haven't had a chance to implement them yet. It's on my radar though!

Finally, even if Tab Stash breaks entirely or you decide to stop using it, you can be assured your tabs are stored safely as bookmarks in Firefox's bookmarks manager. Check the "Tab Stash" folder under "Other Bookmarks". (And yes, you can use the built-in bookmarks manager or any other extension to freely reorganize your Tab Stash bookmarks, although note that Tab Stash only understands a single level of nested folders and will ignore more deeply-nested folders.)

**Comment by joshgold22** [comment]
Ah thanks. I had figured out how to collapse the view, but it still seemed to me like the performance was because of my open tabs rather than my stash bookmarks.  It was very slow (30s-60s?) from the start of trying out the extension, when I didn't have much stashed at all, but did, as always, have too many open tabs.  If you're right about the source of the performance problems, though, then collapsing it should be fine.  (I'd still consider starting out with it collapsed by default, since I'm usually more interested in the stashes themselves, but I can't say what's more typical.)

It's good to know you can't really lose a stashed tab if you know where to find them, but I'd still suggest it's possible to lose work that was in progress in  tabs when stashing by mistake, some authenticated webpages don't reopen easily to their last state, and it can be overwhelming if it's a lot of tabs   Also newer users (including me) may not know that they're protected.  Confirmation dialog with the list and count of tabs when it's more than one?  (See Tree Style Tab / Mutliple Tab Handler for a pretty good example of confirm on close.)  -- (Your and my 2nd paragraph here might relate more to #74 than this issue.)

**Comment by josh-berry** [comment]
hm, 30-60 seconds is way too long to load; that sounds like it needs some investigation. I thought you were talking about a few seconds—I guess I should have asked. :) How many tabs you have open and/or stashed?

Regarding work-in-progress tabs, the default is to hide, not close them, and then unload hidden tabs only after they haven't been touched in a while. So if you restore them within a few minutes or so, whatever was going on inside your tab (authentication, half-written blog posts, etc.) should still be there. (All of this is adjustable in the settings.)

**Comment by josh-berry** [comment]
Just wanted to ping on this one again—for the performance issue, can you share stats on number of tabs/stashes, and maybe memory/CPU usage while trying to open the UI?

**Comment by joshgold22** [comment]
A lot of tabs could be 500.   I'm not sure that's it, though:  I was curious enough to run a profiler while opening the Stashes view in a tab.  The time is mostly spent on hold waiting for garbage collection.  I tried running again with just a couple of tabs open and it took a similar amount of time (still a lot of garbage collection, but looked a little different, coming in short bursts over and over throughout, instead of a few long, sustained delays.  There are interactions with other extensions' listeners getting triggered, so I suspected that it was an addon, but when I disabled a number of the most visible from the profile the overall picture didn't change.  I'd be curious about the allocations, and I was also wondering, is it possible opening the Stashes tab is somehow triggering update events on *other* tabs?  It's just a hypothesis, but that might point to why a number of other extensions are lighting up.

**Comment by josh-berry** [comment]
Sorry it's taken me some time to come back to this.

500 tabs doesn't seem to me to be unreasonable—I've heard of other users with thousands of stashed tabs that don't seem to have this problem. The garbage-collection behavior you're describing lines up with what I've seen in my own profiles—when loading the bookmarks from the browser, the page appears idle (with periodic garbage collections).

The only reason I can think of for events on other extensions being triggered is if you have some other extension which is also monitoring the set of open tabs. Firefox will trigger events on tab open, navigation, close, etc. which any extension can listen to.

One more question—how many _total bookmarks_ do you have (even outside of Tab Stash)? I'm wondering if the load time is correlated not just with the bookmarks in your stash, but the entire bookmark tree.

**Comment by joshgold22** [comment]
I can confirm your intuition was right that the issue is the total number bookmarks, rather than the size of what's in the tab stash bookmark folder.  I know this because I just decided for other reasons to delete almost all of my old bookmarks (I won't estimate, but the number was huge) and I thought of trying out tab stash: Showing the stashes is now no longer slow for me, whether in a new tab or in the sidebar.  (I didn't touch the bookmarks in the tab stash folder.)

**Comment by josh-berry** [comment]
Got it, thanks for the confirmation--that tells me what the culprit is then. Glad to hear things are faster for you now that you've deleted some old bookmarks.

I have a plan to address this by preloading the stash ahead of time, but unfortunately it's not a small fix, since the UI needs to coordinate with the background page (which is always loaded). So it may take some time to make it into a release. But I'm glad we were able to find the root cause at least.

**Comment by josh-berry** [comment]
Small update on this—I think a better plan will be to lazily load only the parts of the stash that can be shown, and avoid loading other parts of the bookmarks tree unless they are needed. This is likely to save memory as well as improve performance.

**Comment by josh-berry** [comment]
The UI in the next release should now become usable much faster after load. It will load the stash (but not the whole bookmark tree) in the background, and only draw the parts of the UI it actually needs to.

**Comment by alphapapa** [comment]
@josh-berry Thanks for your work on this.  I have almost 3,000 tabs stashed at the moment (wow, they add up quickly), and I had noticed that the sidebar was becoming slow to open, to the point that I was beginning to hesitate to use the extension sometimes (and if I accidentally left it open when I closed Firefox, it would significantly slow Firefox's launch next time).  In the new 3.1 version, the sidebar loads so much faster!  Great work!

**Comment by josh-berry** [comment]
@alphapapa My pleasure, really happy to hear that it's helped! 👍

**Comment by MrConorAE** [comment]
Just also wanted to chime in and say that the 3.1 update has completely solved the issues I mentioned in #508. Thank you!

(p.s. it's also very fun to watch the counters in the search box tally up all your tabs the first time you open the sidebar)
